clientconn.go

```diff
@@ -854,176 +850,258 @@ func (ac *addrConn) errorf(format string, a ...interface{}) {
-		if err != nil {
-			cancel()
-			ac.cc.blockingpicker.updateConnectionError(err)
-			ac.mu.Lock()
-			if ac.state == connectivity.Shutdown {
-				// ac.tearDown(...) has been invoked.
-				ac.mu.Unlock()
-				return false, errConnClosing
+		if err := ac.createTransport(newTrID, backoffIdx, addr, copts, backoffDeadline, onGoAway, onClose); err != nil {
+			// errReadTimeOut indicates that the handshake was not received before
+			// the deadline. We exit here because the transport's reader goroutine will
+			// use onClose to reset the transport.
+			if err == errReadTimedOut {
+				return
 			}
-			ac.mu.Unlock()
-			grpclog.Warningf("grpc: addrConn.createTransport failed to connect to %v. Err :%v. Reconnecting...", addr, err)
-			continue
-		}
-		if ac.dopts.waitForHandshake {
+
+			timer := time.NewTimer(backoffDeadline.Sub(time.Now()))
 			select {
-			case <-done:
-			case <-connectCtx.Done():
-				// Didn't receive server preface, must kill this new transport now.
-				grpclog.Warningf("grpc: addrConn.createTransport failed to receive server preface before deadline.")
-				newTr.Close()
-				continue
+			case <-timer.C:
 			case <-ac.ctx.Done():
+				timer.Stop()
+				return
 			}
+
+			continue
 		}
+
+		return
+	}
+}
+
+var errReadTimedOut = errors.New("read timed out")
+
+// createTransport creates a connection to one of the backends in addrs.
+func (ac *addrConn) createTransport(id int32, backoffNum int, addr resolver.Address, copts transport.ConnectOptions, backoffDeadline time.Time, onGoAway func(transport.GoAwayReason), onClose func()) error {
+	timedOutWaitingForPreface := make(chan struct{})
+	// TODO(deklerk): this is unnecessary. In the reader goroutine, we should be able to signal to onClose that the
+	// deadline was exceeded (we can't use a parameter to t.Close because it would mean changes in too many places)
+	onDeadline := func() {
+		close(timedOutWaitingForPreface)
+	}
+
+	target := transport.TargetInfo{
+		Addr:      addr.Addr,
+		Metadata:  addr.Metadata,
+		Authority: ac.cc.authority,
+	}
+	prefaceReceived := make(chan struct{})
+
+	onPrefaceReceipt := func() {
+		// TODO(deklerk): optimization; does anyone else actually use this lock? maybe we can just remove it for this scope
+		ac.mu.Lock()
+		ac.successfulHandshake = true
+		close(prefaceReceived)
+		ac.backoffDeadline = time.Time{}
+		ac.connectDeadline = time.Time{}
+		ac.addrIdx = 0
+		ac.backoffIdx = 0
+		ac.mu.Unlock()
+	}
+
+	ac.mu.Lock()
+	connectDeadline := ac.connectDeadline
+	ac.mu.Unlock()
+
+	// Do not cancel in the success path because of
+	// this issue in Go1.6: https://github.com/golang/go/issues/15078.
+	connectCtx, cancel := context.WithDeadline(ac.ctx, connectDeadline)
+	if channelz.IsOn() {
+		copts.ChannelzParentID = ac.channelzID
+	}
+
+	newTr, err := transport.NewClientTransport(connectCtx, ac.cc.ctx, target, copts, backoffDeadline, onPrefaceReceipt, onDeadline, onGoAway, onClose)
```

Comment:
Why do we need the backoffDeadline passed here?  Why doesn't the transport use the connectCtx's deadline instead?  I think it's not correct to use the backoff deadline except when failures occur and we need to wait before the next iteration.

I think the transport should be using the connect deadline only, and it can read that out of the connectCtx when necessary.

Reply:
You're right. Changed to connectDeadline. Went one step further and removed an unnecessary select that was using backoffDeadline - see https://github.com/grpc/grpc-go/pull/2219/commits/2e3a8d6bad33b216556e502a55431bb25f0be977. Now we should only be using backoffDeadline in nextAddr().